tutorials/rest-http-service-with-axum.mdx

```diff
@@ -0,0 +1,357 @@
+    // that holds the key used to sign cookies
+    key: Key,
+}
+
+// this impl tells `SignedCookieJar` and `PrivateCookieJar how to access the 
+// key from our state
+impl FromRef<AppState> for Key {
+    fn from_ref(state: &AppState) -> Self {
+        state.key.clone()
+    }
+}
+
+[tokio::main]
+async fn main() {
+ 
+let state = AppState {
+    key: Key::generate() 
+};
+
+// ... the rest of your code
+}
+```
+
+Now that we're done with that, we can pass our cookiejar into whatever function we'd like to use it in, and if there's any changes to the cookiejar that we want to pass back to the requesting client then we need to add it as a return type, like so:
+
+```rust
+async fn check_cookie(
+    jar: PrivateCookieJar,
+) -> impl IntoResponse {
+
+    if !jar.get("hello") {
+// a similar thing can be done for deleting cookies, getting cookies etc
+         (jar.add(Cookie::new("hello", "world")), StatusCode::CREATED)
+} 
+
+StatusCode::OK
+}
+```
+
+However, what if we want to make our own cookies? Suppose we have a backend that uses CORS and requires secure, same-site cookies - what would we do? In this case, we can declare a variable as a cookie using Axum's `build` method for the Cookie struct. Let's see what building this might look like:
+
+```rust
+// bear in mind this uses the "time" crate for the duration
+
+let session_id = "Hello world!";
+
+let cookie = Cookie::build("foo", session_id)
+    .secure(true)
+    .same_site(SameSite::Strict)
+    .http_only(true)
+    .path("/")
+    .max_age(Duration::WEEK)
+    .finish();
+```
+
+Then we can put this cookie as the cookie we want to add to the jar as the return instead of `Cookie::new`! 
+
+Deleting a cookie is also pretty much the same as adding a cookie; to be able to pass the changes properly, the deletion of the cookie needs to be specified in the return otherwise it will not delete properly:
+
+```rust
+async fn logout(
+jar: PrivateCookieJar
+) -> impl IntoResponse {
+    (jar.remove(Cookie::named("foo"), StatusCode::OK)
+}
+```
+
+### Middleware
+
+Middleware is essentially a function that runs before the client hits an endpoint; it can be used for things like adding a wait time to prevent server overload, to validating user sessions from cookies. Fortunately, middleware in Axum is quite simple to use! 
+
+There's two ways to be able to create middleware in Axum, but we will be focusing on the more simple way which is to just write a function that uses Axum's middleware utilities as parameters, then simply just adding it to our router and declaring it as middleware. We can also add middleware that uses state this way too, which is great as it means we can share a database connection pool (for example) in the middleware:
+
+```rust
+async fn check_hello_world<B>(
+req: Request<B>,
+next: Next<B>) -> Result<Response, StatusCode> {
+let string = "Hello world!";
+
+if string != "Hello world!" {
+    return Err(StatusCode::INTERNAL_SERVER_ERROR);
+}
+
+Ok(next.run(req).await)
+}
```

Comment:
I think it would be more useful if this example checked something on the request, rather than matching on a hardcoded string.

Reply:
Ok no worries, I'll amend this